### PR TITLE
Detachment of env and proxy; avoiding copies of the environments

### PR DESCRIPTION
**Pull request by alexhernandezgarcia**
This PR refactors several important aspects of the code base:
- Avoid unnecessary copies of environments to sample trajectories.
- Detachment of environment and proxy
- Dropping the convention of range of values in the proxies

## Copies of environments

Each trajectory in a GFlowNet batch is constructed by evolving one environment, either forwards or backwards. So far, to construct each trajectory, a reference environment was copied and reset. This PR replaces this copying operation by storing an "environment maker" as an attribute of the agent (as though it was just the environment class) and instead creating a new environment instance for each trajectory.

## Detachment of environment and proxy

This addresses Issue https://github.com/alexhernandezgarcia/gflownet/issues/288.

So far, the proxy has been part of GFlowNet environments as an attribute. The rationale was to be able to compute the reward of a trajectory directly from the environment, for example with `env.reward()`. While this is a nice feature, it comes at a cost and it has prevented extending the flexibility to use different proxy-to-reward functions, a better handling of the computation of log-rewards for better numerical stability, as well the simple use of alternative baselines to GFlowNets.

In this PR, the proxy and the environments have been detached and become (almost fully) independent of each other. Together with this change, I have also tried to improve the implementation for handling the conversion of proxy values to rewards or log rewards, to make it for efficient, numerically more stable and more flexible to extend.

These are some of the new features:
- Several proxy-to-reward functions are transparently supported: identity, absolute value, power (to β), exponential (Boltzmann), shift (sum of β) and product (with β).
- All this is now implemented in the base Proxy class.
- One can easily create a new custom proxy-to-reward function in their custom proxy class and use it by simply setting it as a callable to in `self.reward_function`.
- Since all losses are computed in log domain, now the log-rewards can directly be computed from the proxy values without first computing rewards and then taking the log. This may be particularly important for numerical stability when using exponential proxy-to-reward functions.

## Dropping convention of range of values in the proxies

So far, we (sort of) followed a convention by which lower proxy values (and negative) were better. This has created a lot of confusion because not all proxies make much sense in this convention. 

With this PR, this convention is dropped altogether and the guiding principle now is that the user is free to create any proxy with any range of values and any direction (maximisation or minimisation), _but_ they are responsible to use an appropriate proxy-to-reward function. This is now much easier given the changes described in the previous section.

## To-do list

Some of these things may be left for a separate PR.

- [x] Sanity check experiments
- [x] IMPORTANT: fix FL loss with exponential proxy2reward (see Tetris) - currently optimises opposite direction
- [x] Fix sanity check experiments with tetris plotting
- [x] Adapt methods in environments (e.g. `sample_from_reward()`, `compute_train_energy_proxy_and_rewards()`...) that still use the old methods in the base env.
- [x] Consider getting rid of `self.reward_norm` and co. in the base env.
- [x] Fix broken common tests with minimal runs.
- [x] Update all configs...
- [x] ~Avoid re-computing proxy values for Buffer~ This is left for a future PR.
- [x] Update outputs of all proxies: the output does not need to be negative anymore, then we can make `identity` the default.
- [x] Write doc in Proxy base about `reward_func` and rest of attributes.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -37,12 +38,16 @@ def main(config):
         float_precision=config.float_precision,
     )
     # The proxy is passed to env and used for computing rewards
-    env = hydra.utils.instantiate(
+    # Using Hydra's partial instantiation, see:
+    # https://hydra.cc/docs/advanced/instantiate_objects/overview/#partial-instantiation
+    env_maker = hydra.utils.instantiate(
         config.env,
         proxy=proxy,
```
Comment (alexhernandezgarcia) [inline main.py:45]: Don't pass proxy to env.

**Comment by alexhernandezgarcia** [comment]
Rethink whether absolute should be the default.

**Comment by alexhernandezgarcia** [comment]
Ok, to me this is now ready to merge. @carriepl @AlexandraVolokhova I'm tagging you as reviewers in case you want to take a quick look and to let you know. Feel free to go ahead and approve and merge unless you have any concerns. Sanity checks look perfect: https://wandb.ai/alexhg/gfn_sanity_checks (the issue with FL was that the sanity checks config files was incorrect for the new proxy - fixed in https://github.com/alexhernandezgarcia/gflownet/pull/299/commits/361cdc0f6ef4268f7f3f1c3046f07428e8539633)

**Comment by alexhernandezgarcia** [comment]
@ginihumer I am tagging you here because this will be merged soon into the main branch and then into the activelearning branch, which will force making changes (for the better) on the active learning repo. No need to check the code details, but it's worth reading the description at least to understand what will change. In a nutshell, that proxies don't need to be negative and the lower the better, but can be anything and can be controlled by `reward_function`.

**Review by carriepl — COMMENTED** [review]
There's a huge amount of work in this PR! Overall, it looks pretty good to me.

File: config/logger/base.yaml [hunk 2]
```diff
@@ -26,6 +26,8 @@ test:
   logprobs_bootstrap_size: 10000
   # Maximum number of test data points to compute log likelihood probs.
   max_data_logprobs: 1e5
+  # Number of points tor obtain a grid to estimate the reward density
+  n_grid: 40401
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply config/logger/base.yaml:30]: - Typo fixed
- I have set the default to 40000 = 200 x 200 so that it's less surprising. It's indeed coming from the Cube experiments but this can be relevant for other cases and I think it make sense to keep the default number to something we found reasonable in the Cube.
Commit: https://github.com/alexhernandezgarcia/gflownet/pull/299/commits/c9879de9ca370d93f6d21b333340c7560c98f98c

File: config/proxy/base.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# - shift
+# - product
+# Alternatively, it can be a callable of the function itself.
+reward_function: identity
+# A callable of the proxy-to-logreward function. 
+# None by default, which takes the log of the proxy-to-reward function
+logreward_function: null
+# Arguments of the proxy-to-reward function.
+# The default functions use an argument with key beta
+reward_function_kwargs: {}
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply config/proxy/base.yaml:17]: Sure, done here: https://github.com/alexhernandezgarcia/gflownet/commit/54ffaf1b1412c0f4b6368549e8c708156bf65eb5

The reward_norm arg is deprecated now.

File: config/experiments/icml23/ctorus.yaml [hunk 4]
```diff
@@ -56,7 +55,7 @@ logger:
     - continuous
     - ctorus
   test:
-    period: 25
+    period: 500
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply config/experiments/icml23/ctorus.yaml:58]: Fair point. I have reverted this and apply the change to the sanity check experiments. Commit: https://github.com/alexhernandezgarcia/gflownet/commit/73a874974ff60e27de9b99a20520c077511b2796

File: gflownet/envs/htorus.py [hunk 5]
```diff
@@ -516,142 +533,192 @@ def get_uniform_terminating_states(
...
-            [samples_mesh, torch.ones(samples_mesh.shape[0], 1)], 1
-        ).to(self.device)
-        rewards = torch2np(
-            self.proxy2reward(self.proxy(self.states2proxy(states_mesh)))
-        )
+        """
+        Plots the reward contour alongside a batch of samples.
+
+        The samples are previously augmented in order to visualise the periodic aspect
+        of the sample space. It is assumed that the samples and the rewards are sorted.
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/envs/htorus.py:580]: I have extended the documentation of this part and change a bit the code to make everything hopefully cleaner. https://github.com/alexhernandezgarcia/gflownet/commit/fa40c51321b39340137d3a08250d7289a60a3f4a

File: gflownet/gflownet.py [hunk 6]
```diff
@@ -1410,6 +1507,60 @@ def test_top_k(self, it, progress=False, gfn_states=None, random_states=None):
...
+            Small epsilon parameter for rejection sampling.
+
+        Returns
+        -------
+        samples_final : list
+            The list of samples drawn from the reward distribution in environment
+            format.
+        """
+        samples_final = []
+        max_reward = self.proxy.proxy2reward(self.proxy.min)
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/gflownet.py:1541]: Oh dear, good catch! This is pretty wrong and it's gonna need a bunch of changes that will delay things a bit...

File: tests/gflownet/proxy/test_base.py [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+
+@pytest.fixture()
+def uniform():
+    return Uniform(device="cpu", float_precision=32)
+
+
+@pytest.fixture()
+def proxy_identity(beta):
+    return Uniform(
+        reward_function="power",
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply tests/gflownet/proxy/test_base.py:18]: Thanks for the catch, fixed here: https://github.com/alexhernandezgarcia/gflownet/commit/a9ef286c91c8c72656f9e246d0e5aff2f5af091b

File: tests/gflownet/proxy/test_base.py [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+    )
+    nan_equal = torch.equal(comp_nan, exp_nan)
+    return notnan_allclose, nan_equal
+
+
+@pytest.mark.parametrize(
+    "beta, proxy_values, rewards_exp, logrewards_exp, logrewards_exp_clipped",
+    [
+        (
+            1,
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply tests/gflownet/proxy/test_base.py:124]: Sure https://github.com/alexhernandezgarcia/gflownet/commit/abe50cba5e07a3f5e5a9a1d30de270981800fcf5

File: tests/gflownet/proxy/test_base.py [hunk 9]
```diff
@@ -0,0 +1,595 @@
...
+)
+def test_reward_function_callable__behaves_as_expected(
+    proxy_callable,
+    reward_function,
+    logreward_function,
+    proxy_values,
+    rewards_exp,
+    logrewards_exp,
+):
+    proxy = proxy_callable
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply tests/gflownet/proxy/test_base.py:587]: I'm pretty sure it works. These arguments are passed to the fixture `proxy_callable` defined in https://github.com/alexhernandezgarcia/gflownet/blob/dont-copy-envs/tests/gflownet/proxy/test_base.py#L66

It's the same as with the argument `beta` in the other tests, which is passed to the fixtures (and is not used explictily within the test).

**Comment by alexhernandezgarcia** [comment]
@carriepl Thank you so much for the review! Would you mind taking a quick look at my comments and resolving yours if the comment is addressed. There should be one remaining issue only, if I am not mistaken.

File: tests/gflownet/proxy/test_base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (carriepl) [reply tests/gflownet/proxy/test_base.py:617]: Ok, I wasn't aware that Pytest fixtures could implicitely use the test parameters. Good to know!

**Comment by carriepl** [comment]
> @carriepl Thank you so much for the review! Would you mind taking a quick look at my comments and resolving yours if the comment is addressed. There should be one remaining issue only, if I am not mistaken.

I just looked. It all looks great. Only that one issue remaining.

**Comment by alexhernandezgarcia** [comment]
Yes this is nice! I use it more and more, for example in the tests of the new Crystal subenvs.

On 21 May 2024 12:59:01 GMT-04:00, carriepl ***@***.***> wrote:
>@carriepl commented on this pull request.
>
>
>
>> +                1.0,
>+                8.1031e03,
>+            ],
>+            [-11, -2, -1.5, -1.1, -1, -0.9, -0.5, 0, 9],
>+        ),
>+    ],
>+)
>+def test_reward_function_callable__behaves_as_expected(
>+    proxy_callable,
>+    reward_function,
>+    logreward_function,
>+    proxy_values,
>+    rewards_exp,
>+    logrewards_exp,
>+):
>+    proxy = proxy_callable
>
>Ok, I wasn't aware that Pytest fixtures could implicitely use the test parameters. Good to know!
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/alexhernandezgarcia/gflownet/pull/299#discussion_r1608660003
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- Sent from /e/ Mail.

File: gflownet/gflownet.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/gflownet.py]: How about this? https://github.com/alexhernandezgarcia/gflownet/pull/299/commits/a1462f1be0ba9df84aefe592278bb42292b571c1

**Review by carriepl — APPROVED** [review]
Looks good to me. Many thanks!
